### PR TITLE
✨ Add validating webhook for AWSMachineTemplate

### LINKED ISSUE #886: Add validating webhooks
/kind feature

**Describe the solution you'd like**
Using validating webhook, we need to add validation logic for AWSMachineTemplate objects to achieve immutability.

### PR DESCRIPTION
**Pull request by tahsinrahman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #886

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -40,5 +40,5 @@ RUN CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} \
 FROM gcr.io/distroless/static:latest
 WORKDIR /
 COPY --from=builder /workspace/manager .
-USER nobody
+# USER nobody
```
Comment (tahsinrahman) [inline Dockerfile]: it is removed as I was getting this error
```
E0910 19:34:39.892550       1 main.go:164] setup "msg"="problem running manager" "error"="listen tcp :443: bind: permission denied"  
```

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha2-awsmachinetemplate,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=awsmachinetemplates,versions=v1alpha2,name=vawsmachinetemplate.kb.io
+
+var _ webhook.Validator = &AWSMachineTemplate{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachineTemplate) ValidateCreate() error {
+	return r.validateAWSMachineTemplate()
```
Comment (ncdc) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: I would `return nil` here, and add a comment that there is no validation on create.

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	oldAWSMachineTemplate := old.(*AWSMachineTemplate)
+	return validateUpdate(&r.Spec.Template.Spec, &oldAWSMachineTemplate.Spec.Template.Spec)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachineTemplate) ValidateDelete() error {
+	return nil
+}
+
+func (r *AWSMachineTemplate) validateAWSMachineTemplate() error {
```
Comment (ncdc) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: I'd remove this

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	return nil
+}
+
+// validateUpdate checks that no immutable fields have been updated
+func validateUpdate(spec *AWSMachineSpec, oldSpec *AWSMachineSpec) error {
+	// Instance Type
+	if spec.InstanceType != oldSpec.InstanceType {
+		return errors.Errorf("instance type cannot be mutated from %q to %q", oldSpec.InstanceType, spec.InstanceType)
```
Comment (ncdc) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: It would probably be better to aggregate these using k8s.io/apimachinery/pkg/api/errors.NewInvalid (or whatever makes the most sense). I would aggregate each field-specific error into a `field.ErrorList`, using methods such as `field.Forbidden(field.NewPath("spec", "instanceType"), "instance type is immutable")`

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	// as a *string, so do the same here.
+	if spec.Subnet != nil {
+		if *spec.Subnet.ID != *oldSpec.Subnet.ID {
+			return errors.Errorf("machine subnet ID cannot be mutated from %q to %q",
+				*oldSpec.Subnet.ID, *spec.Subnet.ID)
+		}
+	}
+
+	// Public IP
+	if *spec.PublicIP != *oldSpec.PublicIP {
```
Comment (ncdc) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: nil check?

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	if spec.RootDeviceSize > 0 && spec.RootDeviceSize != oldSpec.RootDeviceSize {
+		return errors.Errorf("Root volume size cannot be mutated from %v to %v", oldSpec.RootDeviceSize, spec.RootDeviceSize)
+	}
+
+	// Subnet ID
+	// spec.Subnet is a *AWSResourceReference and could technically be
+	// a *string, ARN or Filter. However, elsewhere in the code it is only used
+	// as a *string, so do the same here.
+	if spec.Subnet != nil {
+		if *spec.Subnet.ID != *oldSpec.Subnet.ID {
```
Comment (ncdc) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: nil check on oldSpec.Subnet?

File: api/v1alpha2/zz_generated.deepcopy.go [hunk 7]
```diff
@@ -22,7 +22,7 @@ package v1alpha2
 
 import (
 	"k8s.io/api/core/v1"
-	runtime "k8s.io/apimachinery/pkg/runtime"
```
Comment (ncdc) [inline api/v1alpha2/zz_generated.deepcopy.go]: Did you regenerate this file or what led to this change?

File: config/certmanager/certificate.yaml [hunk 8]
```diff
@@ -1,23 +1,23 @@
 # The following manifests contain a self-signed issuer CR and a certificate CR.
 # More document can be found at https://docs.cert-manager.io
-apiVersion: certmanager.k8s.io/v1alpha2
+apiVersion: certmanager.k8s.io/v1alpha1
```
Comment (ncdc) [inline config/certmanager/certificate.yaml]: Why the version downgrade?

File: config/default/kustomization.yaml [hunk 9]
```diff
@@ -39,9 +39,39 @@ patchesStrategicMerge:
...
-#- manager_webhook_patch.yaml
+- manager_webhook_patch.yaml
 
 # [CAINJECTION] Uncomment next line to enable the CA injection in the admission webhooks.
 # Uncomment 'CAINJECTION' in crd/kustomization.yaml to enable the CA injection in the admission webhooks.
 # 'CERTMANAGER' needs to be enabled to use ca injection
-#- webhookcainjection_patch.yaml
+- webhookcainjection_patch.yaml
+
+# the following config is for teaching kustomize how to do var substitution
```
Comment (ncdc) [inline config/default/kustomization.yaml]: Why did you move this?

File: config/webhook/kustomizeconfig.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# the following config is for teaching kustomize where to look at when substituting vars.
+# It requires kustomize v2.1.0 or newer to work properly.
```
Comment (ncdc) [inline config/webhook/kustomizeconfig.yaml]: We require v3.1.0 or newer

File: Dockerfile (hunk 1 above)
Comment (ncdc) [reply Dockerfile]: Let's undo this. We can control it using the manager Options struct, field=Port. We can create a flag for this in main.go.

File: config/certmanager/certificate.yaml (hunk 8 above)
Comment (tahsinrahman) [reply config/certmanager/certificate.yaml]: certmanager is actually in v1alpha1 https://github.com/jetstack/cert-manager/blob/master/deploy/manifests/00-crds.yaml

File: config/webhook/kustomizeconfig.yaml (hunk 10 above)
Comment (tahsinrahman) [reply config/webhook/kustomizeconfig.yaml]: I created a sample project using kubebuilder 2.0 and copied the file from there

File: config/default/kustomization.yaml (hunk 9 above)
Comment (tahsinrahman) [reply config/default/kustomization.yaml]: I was getting this error
```console
$ kustomize build config/default/
Error: var '{SERVICE_NAME ~G_v1_Service {metadata.name}}' cannot be mapped to a field in the set of known resources
```

besides, kubebuilder creates it inside this directory, so i moved it there

File: api/v1alpha2/zz_generated.deepcopy.go (hunk 7 above)
Comment (tahsinrahman) [reply api/v1alpha2/zz_generated.deepcopy.go]: oops, probably my IDE gofmt/goimport-ed this file

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachineTemplate) ValidateDelete() error {
+	// there's no validation for AWSMachineTempalate on delete
+	return nil
+}
+
+// validateUpdate checks that no immutable fields have been updated
+func validateUpdate(spec *AWSMachineSpec, oldSpec *AWSMachineSpec) error {
```
Comment (detiber) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: Are there any fields that we want to be mutable in the template types? @ncdc @vincepri I thought we had wanted to avoid allowing mutability of the template types and instead wanted to require users to update the reference in the MachineDeployment/MachineSet?

If we do want to force full immutability, then we can likely create a generic webhook in CAPI that blocks any update, and just register the webhook for the AWSMachineTemplate type here in CAPA.

File: config/default/kustomization.yaml (hunk 9 above)
Comment (ncdc) [reply config/default/kustomization.yaml]: Ah, thanks. Was this maybe generated by an earlier version of kubebuilder, perhaps? I'm definitely 👍 to match how/where kubebuilder generates things.

File: api/v1alpha2/awsmachinetemplate_webhook.go (hunk 11 above)
Comment (ncdc) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: We can allow updates to metadata, but probably all other updates should be rejected. I think we could probably write some generic validation helpers in CAPI, but afaik we still need to do the full webhook code generation & registration for each type.

Comment (detiber) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: According to the docs, we can configure which resources are subject to which admission webhooks at runtime: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#configure-admission-webhooks-on-the-fly

Comment (vincepri) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: We should avoid naming changes to resources given that they're references elsewhere, unless the OwnerReferences get updated automagically?

Comment (detiber) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @ncdc I just want to avoid us requiring  all providers to need to re-implement the exact same thing

File: Dockerfile (hunk 1 above)
Comment (joonas) [reply Dockerfile]: Filed this as https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1120

File: api/v1alpha2/awsmachinetemplate_webhook.go (hunk 11 above)
Comment (vincepri) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: I think that's fair, +1

Comment (ncdc) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @detiber are you thinking that there would be one `ValidatingWebhookConfiguration` in CAPI and we would have documentation for how people installing providers would need to update it to include their resources in the generic template validator?

Comment (detiber) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @ncdc possibly, we could also potentially give RBAC permissions to providers to register the types in the ValidatingWebhookConfiguration themselves when the controller manager starts up.

Comment (vincepri) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: Can we dynamically watch unstructured objects in a group? And maybe discover those groups automatically?

CAPI should already have permissions for bootstrap.* and infrastructure.* groups

Comment (ncdc) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: We typically advise that registering things like CRDs and webhook configurations are tasks better left to admin accounts. I would be hesitant to let the controllers do this.

Comment (ncdc) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: This sounds like something that clusteradm could handle long-term

Comment (vincepri) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: Maybe we're talking about different things, I was thinking to make this Webhook server available by default in CAPI, and give it permissions to the groups and validate those objects

Comment (ncdc) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @vincepri yes, we are talking about different things. Thanks for pointing that out. So you'd have this generic webhook use API discovery and look for `*Template` types? Although that wouldn't automatically cover third party providers whose API groups are different.

Comment (vincepri) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: As long as it's documented, I think that should be fine. Technically, the same applies for today's default RBAC permissions in CAPI, where we assume `bootstrap.<>` and `infrastructure.<>` as default

Comment (detiber) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @ncdc agreed that clusteradm would be the best place long term to handle this

Comment (detiber) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @vincepri @ncdc the validating webhook requests are initiated by the API Server, I'm not sure there is a way to use API discovery to register types to watch at runtime as we do with the controllers.

Comment (ncdc) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @detiber that's right. So the only way to handle this is with the details in the ValidatingWebhookConfiguration.

What do you all think about putting the generic *Template validation as a helper function in cluster-api, but at least for now, having each provider implement their own webhooks for their own templates (using the generic function from cluster-api)?

Comment (detiber) [reply api/v1alpha2/awsmachinetemplate_webhook.go]: @ncdc I think that's fine as long as we track improving that experience later when clusteradm is available.

File: api/v1alpha2/awsmachinetemplate_webhook.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+	if spec.RootDeviceSize > 0 && spec.RootDeviceSize != oldSpec.RootDeviceSize {
+		errs = append(errs, field.Forbidden(field.NewPath("spec", "rootDeviceSize"),
+			"root volume size is immutable"))
+	}
+
+	// Subnet ID
+	// spec.Subnet is a *AWSResourceReference and could technically be
+	// a *string, ARN or Filter. However, elsewhere in the code it is only used
+	// as a *string, so do the same here.
+	if aws.StringValue(spec.Subnet.ID) != aws.StringValue(oldSpec.Subnet.ID) {
```
Comment (liztio) [inline api/v1alpha2/awsmachinetemplate_webhook.go]: We should probably check that `spec.Subnet` isn't nil

File: config/webhook/manifests.yaml [hunk 13]
```diff
@@ -0,0 +1,26 @@
+
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  creationTimestamp: null
+  name: validating-webhook-configuration
+webhooks:
+- clientConfig:
+    caBundle: Cg==
```
Comment (liztio) [inline config/webhook/manifests.yaml]: What does this do?

Comment (ncdc) [reply config/webhook/manifests.yaml]: This is generated by kubebuilder. I guess it's a required field. Maybe cert-manager or something else updates it? It decodes to `\n`.

File: config/webhook/service.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: webhook-service
```
[the remark this answers is not in the document]
Comment (tahsinrahman) [reply config/webhook/service.yaml]: yes, kustomize adds `capa-` prefix to name and namespace!

File: controllers/awsmachine_controller.go [hunk 15]
```diff
@@ -282,13 +280,6 @@ func (r *AWSMachineReconciler) reconcileNormal(ctx context.Context, machineScope
 		return reconcile.Result{}, nil
 	}
 
-	// TODO(ncdc): move this validation logic into a validating webhook
```
Comment (chuckha) [inline controllers/awsmachine_controller.go]: I don't see this logic moved anywhere, am i misunderstanding this Webhook?

Comment (ncdc) [reply controllers/awsmachine_controller.go]: It was in an earlier version of this PR. Not sure why it got deleted?

Comment (tahsinrahman) [reply controllers/awsmachine_controller.go]: as entire spec is immutable, instead of checking every field of spec, i did this
```
if !reflect.DeepEqual(newAWSMachineTemplate.Spec, oldAWSMachineTemplate.Spec) {
	return errors.New("awsMachineTemplateSpec is immutable")
}
```

Comment (detiber) [reply controllers/awsmachine_controller.go]: This validation was being used by the AWSMachine controller, where the current webhook is targeting AWSMachineTemplate.

Comment (tahsinrahman) [reply controllers/awsmachine_controller.go]: got it!
that means, we need to deploy another validating webhook for `AWSMachine` types, to move these logic, right?

Comment (detiber) [reply controllers/awsmachine_controller.go]: correct, thanks :)

Comment (ncdc) [reply controllers/awsmachine_controller.go]: @tahsinrahman are you going to add the AWSMachine validation webhook in this PR?

File: config/crd/kustomization.yaml [hunk 16]
```diff
@@ -9,7 +7,7 @@ resources:
 - bases/infrastructure.cluster.x-k8s.io_awsmachinetemplates.yaml
 # +kubebuilder:scaffold:crdkustomizeresource
 
-#patches:
+patchesStrategicMerge:
```
[the remark this answers is not in the document]
Comment (tahsinrahman) [reply config/crd/kustomization.yaml]: no, they're for enabling conversion webhooks

```
# [WEBHOOK] To enable webhook, uncomment all the sections with [WEBHOOK] prefix.
# patches here are for enabling the conversion webhook for each CRD
```

[the remark this answers is not in the document]
Comment (ncdc) [reply config/crd/kustomization.yaml]: Thanks. It's weird that conversions are enabled in config/crd, and mutating/validating are in config/default.

File: controllers/awsmachine_controller.go (hunk 15 above)
Comment (tahsinrahman) [reply controllers/awsmachine_controller.go]: Sure, I can add it here, or if it becomes too big for review, i can do it in separate pr, which one works best for you? :smile:

Comment (ncdc) [reply controllers/awsmachine_controller.go]: Either way is fine. If you keep them separate, you can retitle this PR to "Add validating webhook for AWSMachineTemplate"?

File: config/webhook/manifests.yaml [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    service:
+      name: webhook-service
+      namespace: system
+      path: /validate-infrastructure-cluster-x-k8s-io-v1alpha2-awsmachinetemplate
+  failurePolicy: Fail
+  name: vawsmachinetemplate.kb.io
```
Comment (vincepri) [inline config/webhook/manifests.yaml]: How was this name generated?

Comment (ncdc) [reply config/webhook/manifests.yaml]: kubebuilder does it this way

Comment (vincepri) [reply config/webhook/manifests.yaml]: Got it, was mostly curious as it reads funny

**Comment by ncdc** [comment]
> Do we need to cut release-0.4 first?

No, we can branch from any commit at any time.

**Comment by tahsinrahman** [comment]
Do we want to merge this now? It'll break clusterctl

**Comment by vincepri** [comment]
Let's hold on to this until we have all the others open PR merged in 0.4 and we cut the release branch

**Comment by ncdc** [comment]
We also don't have to rush this in. Let's think more about the "clusterctl on master" impact.

**Comment by ncdc** [comment]
I'm thinking this will need to move into the v1alpha3 api package (#1202), right?

**Comment by tahsinrahman** [comment]
I updated the pr with v1alpha3 package!

**Comment by detiber** [comment]
/lgtm
I believe the hold can be lifted now that we have the release-0.4 branch as well. @ncdc?

**Comment by ncdc** [comment]
Yes, master is open for v1alpha3.
/hold cancel
